pypeit/core/skysub.py

```diff
@@ -675,29 +675,75 @@ def local_skysub_extract(sciimg, sciivar, tilts, waveimg, global_sky, rn2_img,
     # TODO Can this be simply replaced with spat_img above (but not spat_pix since that could have holes)
     spatial_img = thismask * ximg * (np.outer(xsize, np.ones(nspat)))
 
-    # Loop over objects and group them
-    i1 = 0
-    while i1 < nobj:
-        group = np.array([], dtype=np.int)
-        group = np.append(group, i1)
-        # The default value of maskwidth = 3.0 * FWHM = 7.05 * sigma in objfind with a log(S/N) correction for bright objects
-        min_spat1 = np.maximum(sobjs[i1].TRACE_SPAT - sobjs[i1].maskwidth - 1, slit_left)
-        max_spat1 = np.minimum(sobjs[i1].TRACE_SPAT + sobjs[i1].maskwidth + 1, slit_righ)
-        for i2 in range(i1 + 1, nobj):
-            left_edge = sobjs[i2].TRACE_SPAT - sobjs[i2].maskwidth - 1
-            righ_edge = sobjs[i2].TRACE_SPAT + sobjs[i2].maskwidth + 1
-            touch = (left_edge < max_spat1) & (sobjs[i2].TRACE_SPAT > slit_left) & (righ_edge > min_spat1)
+    # First, we find all groups of objects to local skysubtract together
```

Comment:
I suggest that we make from line  679 to 731 a method in the specobjs class, which returns the object groups, i.e. 

groups = sobjs.get_extraction_groups()

Or something like that. The reason why I think it should be a method in specobjs is that it only has to do with the specobjs properties, i.e. the TRACE_SPAT, nobjs, and maskwidth. 

The only slight subtlety is when model_full_slit is set above. This argument is used for narrow slit echelle reductions where the goal is to do simply perform local sky-subtraction on the whole slit grouping all objects together. This is being hacked in above at present by just making the maskwidth large such that everything touches. I can live with that remaining as is for now, but it is admittedly not very elegant.

Reply:
The main bit of this code now lives in `SpecObjs.get_extraction_groups`, and I put the DFS routine into `utils` since it could be more widely useful than just here.

For `model_full_slit`, we would want one group with all of the objects, and then the `min_spat` and `max_spat` later just being the slit edges, right?

Reply:
Yeah that is right, one group with all the objects and  min_spat1  and max_spat1 being the slit edges. That said it needs to be the minimum slit edge, but you also need to make sure that min_spat1 and max_spat1 are on the detector, since I think it would crash if you have a negative value or a value greater than nspat. 

I think something like

groups = sobjs.get_extraction_groups(model_full_slit=model_full_slit)

Would be cleaner than the way it is implemented currently. We will need to test all of this though.